File: gradle-plugin/src/main/java/juuxel/ripple/gradle/RippleExtension.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package juuxel.ripple.gradle;

import blue.endless.jankson.Jankson;
import blue.endless.jankson.JsonObject;
import blue.endless.jankson.api.SyntaxError;
import juuxel.ripple.gradle.impl.ProcessedDependency;
import juuxel.ripple.gradle.impl.RippleExtensionImpl;
import juuxel.ripple.processor.NameProcessor;
import juuxel.ripple.processor.NameProcessorIo;
import org.gradle.api.Action;
import org.gradle.api.GradleException;
import org.gradle.api.artifacts.Dependency;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public abstract class RippleExtension {
    private static final Action<DependencySpec> EMPTY_ACTION = spec -> {};

    private List<NameProcessor<?>> processors = new ArrayList<>();

    public List<NameProcessor<?>> getProcessors() {
        return processors;
    }

    public void setProcessors(List<NameProcessor<?>> processors) {
        this.processors = Objects.requireNonNull(processors, "processors");
    }

    /**
     * Adds a processor to Ripple.
     *
     * <p>If the processor parameter is a {@link NameProcessor}, it will be attached directly.
     * Otherwise, it'll be resolved as a {@link File} and parsed using {@link NameProcessorIo}.
     *
     * @param processor the processor or the file path to a processor JSON file
     */
    public void processor(Object processor) {
        if (processor instanceof NameProcessor<?>) {
            processors.add((NameProcessor<?>) processor);
        } else {
            try {
                JsonObject json = Jankson.builder().build().load(asImpl().fileResolver.apply(processor));
                NameProcessorIo.readAll(json).forEach(processors::add);
            } catch (IOException e) {
                throw new GradleException("Could not load name processor config file " + processor, e);
            } catch (SyntaxError e) {
                throw new GradleException("Malformed name processor config file " + processor + ": " + e.getCompleteMessage(), e);
            }
        }
    }

    public Dependency process(Dependency dependency, String tag, Action<DependencySpec> action) {
        return new ProcessedDependency(asImpl(), dependency, tag, processedSpecOf(dependency, tag, action));
    }

    public Dependency process(Dependency dependency, String tag) {
        return process(dependency, tag, EMPTY_ACTION);
    }

    public Dependency process(String dependencyNotation, String tag, Action<DependencySpec> action) {
        return process(asImpl().dependencyCreator.apply(dependencyNotation), tag, action);
    }

    public Dependency process(String dependencyNotation, String tag) {
        return process(dependencyNotation, tag, EMPTY_ACTION);
    }

    private static DependencySpec processedSpecOf(Dependency dependency, String tag, Action<DependencySpec> action) {
        DependencySpec spec = new DependencySpec(
            "ripple." + dependency.getGroup(),
            dependency.getName(),
            dependency.getVersion() + '-' + tag
        );
        action.execute(spec);
        return spec;
    }

    private RippleExtensionImpl asImpl() {
        return (RippleExtensionImpl) this;
    }
}

File: gradle-plugin/src/main/kotlin/juuxel/ripple/gradle/RippleExtension.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package juuxel.ripple.gradle

import blue.endless.jankson.Jankson
import juuxel.ripple.gradle.dependency.DependencySpec
import juuxel.ripple.gradle.dependency.ProcessedDependency
import juuxel.ripple.processor.NameProcessor
import juuxel.ripple.processor.NameProcessorIo
import org.gradle.api.Action
import org.gradle.api.artifacts.Configuration
import org.gradle.api.artifacts.Dependency
import org.gradle.api.file.FileCollection
import java.io.File
import java.nio.file.Path
import kotlin.streams.asSequence

/**
 * The DSL extension of Ripple.
 *
 * Dependency processors can be added with [processor], and dependencies can be processed using [processed] overloads.
 */
open class RippleExtension {
    internal lateinit var detachedConfigGetter: (Dependency) -> Configuration
    internal lateinit var cache: Path
    internal lateinit var fileCollectionGetter: (Set<File>) -> FileCollection
    internal lateinit var fileResolver: (Any) -> File
    internal lateinit var dependencyCreator: (Any) -> Dependency

    internal val processors: MutableList<NameProcessor<*>> = ArrayList()

    /**
     * Adds a [processor] to Ripple.
     *
     * If the [processor] parameter is a [NameProcessor], it will be attached directly.
     * Otherwise, it'll be resolved as a [File] and parsed using [NameProcessorIo].
     */
    fun processor(processor: Any) {
        if (processor is NameProcessor<*>) {
            processors += processor
        } else {
            val json = Jankson.builder().build().load(fileResolver(processor))
            processors += NameProcessorIo.readAll(json).asSequence()
        }
    }

    /**
     * Creates a processed mapping dependency from the [dependencyNotation] with the version [tag].
     * The [tag] should uniquely name the version of the name processors.
     *
     * The optional [specAction] can be used to modify the coordinates of the processed dependency.
     */
    @JvmOverloads
    fun processed(dependencyNotation: String, tag: String, specAction: Action<DependencySpec> = EMPTY_ACTION): Dependency =
        processed(dependencyCreator(dependencyNotation), tag, specAction)

    /**
     * Creates a processed mapping dependency from the [dependency] with the version [tag].
     * The [tag] should uniquely name the version of the name processors.
     *
     * The optional [specAction] can be used to modify the coordinates of the processed dependency.
     */
    @JvmOverloads
    fun processed(dependency: Dependency, tag: String, specAction: Action<DependencySpec> = EMPTY_ACTION): Dependency =
        ProcessedDependency(this, dependency, tag, processedSpecOf(dependency, tag).apply(specAction::execute))

    private fun processedSpecOf(dependency: Dependency, tag: String): DependencySpec =
        DependencySpec(
            "ripple.processed",
            "${dependency.group?.replace('.', '-')}-${dependency.name}",
            "${dependency.version}-$tag"
        )

    companion object {
        private val EMPTY_ACTION: Action<DependencySpec> = Action {}
    }
}
